diff --git a/libs/pmt/common/PMT.cpp b/libs/pmt/common/PMT.cpp
index 5b8a6abc9c..3368d111ac 100644
--- a/libs/pmt/common/PMT.cpp
+++ b/libs/pmt/common/PMT.cpp
@@ -24,8 +24,6 @@ bool isNumber(const std::string &s) {
 namespace pmt {
 
 PMT::~PMT() {
-  StopDump();
-  StopThread();
 };
 
 double PMT::seconds(const Timestamp &timestamp) {
@@ -79,39 +77,6 @@ float State::watts(int i) {
   return watt_[i];
 }
 
-void PMT::StartThread() {
-  SetMeasurementInterval();
-
-  thread_ = std::thread([&] {
-    State state_previous = GetState();
-    assert(state_previous.nr_measurements_ > 0);
-    state_latest_ = state_previous;
-
-    if (dump_file_) {
-      DumpHeader(state_previous);
-    }
-
-    while (!thread_stop_) {
-      std::this_thread::sleep_for(
-          std::chrono::milliseconds(GetMeasurementInterval()));
-      state_latest_ = GetState();
-
-      if (dump_file_ &&
-          (1e3 * seconds(state_previous, state_latest_)) > GetDumpInterval()) {
-        Dump(state_latest_);
-        state_previous = state_latest_;
-      }
-    }
-  });
-}
-
-void PMT::StopThread() {
-  thread_stop_ = true;
-  if (thread_.joinable()) {
-    thread_.join();
-  }
-}
-
 void PMT::StartDump(const char *filename) {
   const char *filename_ = std::getenv(kDumpFilenameVariable.c_str());
   if (filename_) {
@@ -187,13 +152,8 @@ void PMT::SetMeasurementInterval(unsigned int milliseconds) {
 Timestamp PMT::GetTime() { return std::chrono::system_clock::now(); }
 
 State PMT::Read() {
-  const int measurement_interval = GetMeasurementInterval();
-  if (!thread_started_) {
-    StartThread();
-    thread_started_ = true;
-    std::this_thread::sleep_for(
-        std::chrono::milliseconds(measurement_interval));
-  }
+  // Get the latest measurement
+  state_latest_ = GetState();
   return state_latest_;
 }
 
diff --git a/libs/pmt/common/PMT.h b/libs/pmt/common/PMT.h
index 269d54931b..19028b8860 100644
--- a/libs/pmt/common/PMT.h
+++ b/libs/pmt/common/PMT.h
@@ -85,20 +85,11 @@ class PMT {
   Timestamp GetTime();
 
  private:
-  unsigned int measurement_interval_ = 100;  // milliseconds
+  unsigned int measurement_interval_ = 1;  // milliseconds
 
   // The last state set by the thread
   State state_latest_;
 
-  // This thread continuously call GetState to update state_latest_. It is
-  // started automatically upon the first Read() call.
-  std::thread thread_;
-  volatile bool thread_started_ = false;
-  volatile bool thread_stop_ = false;
-
-  void StartThread();
-  void StopThread();
-
   void DumpHeader(const State &state);
 
  protected:
diff --git a/libs/pmt/common/io.h b/libs/pmt/common/io.h
index 2794e2a019..2a6702a9c4 100644
--- a/libs/pmt/common/io.h
+++ b/libs/pmt/common/io.h
@@ -82,7 +82,7 @@ inline size_t pread(int fd, const std::span<std::byte>& byte,
                 static_cast<::off_t>(offset));
     if (data_read < 0) {
       const int errcode = take_and_reset_errno();
-      if (errno == EINTR) {
+      if (errcode == EINTR) {
         // interrupted system call
         continue;
       }
diff --git a/libs/pmt/rapl/RaplImpl.cpp b/libs/pmt/rapl/RaplImpl.cpp
index 369502c103..4f9970e694 100644
--- a/libs/pmt/rapl/RaplImpl.cpp
+++ b/libs/pmt/rapl/RaplImpl.cpp
@@ -115,6 +115,7 @@ void RaplImpl::Init() {
         if (e.code().value() == EACCES) {
           std::cerr << "Please check the permission or try to run as 'root'"
                     << std::endl;
+          std::abort();
         }
       }
     }
@@ -131,10 +132,11 @@ void RaplImpl::Init() {
       uj_offset_[i] = 0;
     }
 
-  } catch (std::exception& e) {
+  } catch (const std::runtime_error& e) {
     std::stringstream message;
     message << "Unable to init rapl plugin: " << e.what();
     std::cerr << message.str() << std::endl;
+    std::abort();
   }
 }
 
diff --git a/libs/pmt/rapl/RaplImpl.h b/libs/pmt/rapl/RaplImpl.h
index a13e32d03f..fb0bf256f2 100644
--- a/libs/pmt/rapl/RaplImpl.h
+++ b/libs/pmt/rapl/RaplImpl.h
@@ -5,6 +5,7 @@
 #include <mutex>
 #include <string>
 #include <vector>
+#include <array>
 
 #include "Rapl.h"
 #include "common/PMT.h"
